### PR TITLE
Print list of missing target features when calling a function with target features outside an unsafe block

### LINKED ISSUE #108680: target_feature 1.1 error should print the list of missing target features
Given the following code:

```Rust
#![feature(target_feature_11)]

#[target_feature(enable = "sse2", enable = "avx", enable = "sse")]
pub fn foo() {}

#[target_feature(enable = "sse")]
fn bar() {
    foo();
}
```

The current output is:

```
error[E0133]: call to function with `#[target_feature]` is unsafe and requires unsafe function or block
 --> src/main.rs:8:5
  |
8 |     foo();
  |     ^^^^^ call to function with `#[target_feature]`
  |
  = note: can only be called if the required target features are available
```

The note is easy to be misinterpreted: the function can also be called if the target features are not available, but it requires an `unsafe` annotation.

I think it would be nice for the note to list the missing target features. In this case, this would be `sse2` and `avx`.
If one or more of the missing target features are enabled in the current build config, like via `-C target-feature` or `-C target-cpu`, or because the default is just that, then it should acknowledge this situation and explain that a `#[target_feature]` [is still required](https://github.com/rust-lang/rust/pull/99767#issuecomment-1447599807). 

So I'd imagine something like:

```
error[E0133]: call to function with `#[target_feature]` is unsafe and requires unsafe function or block
 --> src/main.rs:8:5
  |
8 |     foo();
  |     ^^^^^ call to function with `#[target_feature]`
  |
  = help: in order for the call to be safe, the context requires the following additional target features: sse2, avx
  = note: the sse2 target feature being enabled in the build configuration does not remove the requirement to list it in `#[target_feature]`
```

------

TLDR: I ask for two things:

1. removing the `note` in the general case (I think what it contains is already said in the main error msg), but if there are missing target features that are enabled in the build configuration, emit a note to point out that even though they are enabled, this has no influence on `#[target_feature]` (it's two different systems; see my example for the proposed wording)
1. adding a `help` that lists the missing target features (ideally if too many like more than 5 are missing, then cut the list to keep the error message short ("avx, sse2, sse, aes, foobar, and 3 more"))

cc #69098

@rustbot label A-diagnostics F-target_feature_11

### PR DESCRIPTION
**Pull request by eduardosm**
Fixes https://github.com/rust-lang/rust/issues/108680

Supersedes https://github.com/rust-lang/rust/pull/109710. I used the same wording for the messages, but the implementation is different.

r? @est31

### CODE REVIEW

File: compiler/rustc_middle/src/mir/query.rs [hunk 1]
```diff
@@ -39,10 +39,17 @@ pub enum UnsafetyViolationDetails {
     AccessToUnionField,
     MutationOfLayoutConstrainedField,
     BorrowOfLayoutConstrainedField,
-    CallToFunctionWith,
+    CallToFunctionWith {
+        /// Target features enabled in callee's `#[target_feature]` but missing in
+        /// caller's `#[target_feature]`
```
Comment (kw-fn) [inline compiler/rustc_middle/src/mir/query.rs:44]:
```suggestion
        /// caller's `#[target_feature]`.
```

File: compiler/rustc_mir_transform/src/check_unsafety.rs [hunk 2]
```diff
@@ -367,9 +368,22 @@ impl<'tcx> UnsafetyChecker<'_, 'tcx> {
 
         // Is `callee_features` a subset of `calling_features`?
```
Comment (kw-fn) [inline compiler/rustc_mir_transform/src/check_unsafety.rs:369]:
```suggestion
        // Checks if `callee_features` is a subset of `calling_features`.
```

**Review by est31 — APPROVED** [review]
Thanks for filing!

LGTM except one question: could you add something to the existing test that excercises the `mir_transform_target_feature_call_note` case with multiple target features? Because right now, there is only tests for one.

**Comment by eduardosm** [comment]
Done.

I just learned about `-Z thir-unsafeck`. If I understood it correctly, it is an alternative unsafety checker that runs on a different stage. Do we want to make the change there too?

**Comment by est31** [comment]
> Do we want to make the change there too?

Yeah good point, we'd want that change be present there, too. So the .thir.stderr file should have the same changes in the end as .mir.stderr, ideally.

**Comment by eduardosm** [comment]
Done, also added a test to cover the case of an unsafe function with a `unsafe_op_in_unsafe_fn`  lint.

**Review by est31 — APPROVED** [review]
Great work, thank you!

**Pull request merged**